### PR TITLE
EWS spraying module, Jitter command line option and Dockerfile

### PR DESCRIPTION
**Pull request by puzzlepeaches**
## What?

This pull request includes the following:

- Small updates to existing code to fix some issues introduced following the SMB spraying module addition
- A simple module for password spraying EWS endpoints
- Additions to the requirements.txt file to reflect new libraries used in the EWS spraying module
- The ability to add a jitter time between login attempts for all modules
- A Dockerfile to provide an additional installation option
- Updates to the readme to reflect changes and some other installation options

## Why?

- I needed this for an upcoming engagement :)
- This is one of my favorite tools right now
- To make some stuff easier for new users

## How?

- The EWS spraying module uses requests_ntlm library to issue NTLM authentication attempts over HTTP
- The jitter option allows the user to define a random time variation between login requests 
- The Dockerfile uses the base python:3 container

## Testing?

- I have tried to test the functionality of both big additions to the best of my ability
- I was able to verify the EWS spraying module is working properly across multiple clients and different versions of Exchange
- I have not encountered any odd error cases but I think you should give it a try before pushing to master
- The Dockerfile builds successfully and I have used the pipenv install options several times since I started using this project

## Anything Else?

I am currently working on implementing the following

- The ability for the user to specify the NTLM authentication endpoint as opposed to targeting only EWS
- Configuration file parsing using the ConfigArgParse library as my spraycharles commands are getting pretty long at this point :)
- Notifications using the notifiers library and a configuration file. With both, users wont need to specify their API keys on the command line and can specify multiple notification providers in the config file for team members during ops
- A wiki to help new users with the tool

Please reach out if you want to collaborate on any of the above or have questions about the pull request! Would love to work with you on continuing to improve the tool.

### CODE REVIEW

**Comment by Tw1sm** [comment]
@puzzlepeaches These look like great additions! I will work on testing the docker, EWS and jitter features and bug fixes. Bit busy at the moment, but I wanted to drop in here an acknowledge the pull request and let you know that it's on my to-do list.

I think your roadmap ideas look good as well and I'm down to collaborate as we see fit. I can definitely start a wiki  sometime soon - it would be good to have more documentation on top of the README for the growing pool of arguments and also some documentation on easily building new modules.

File: spraycharles.py [hunk 1]
```diff
@@ -33,6 +34,8 @@ def args():
     parser.add_argument("-f", "--fireprox", type=str, dest="fireprox", help="the url of the fireprox interface, if you are using fireprox.", required=False)
     parser.add_argument("-d", "--domain", type=str, dest="domain", help="HTTP: Prepend DOMAIN\\ to usernames. SMB: Supply domain for smb connection", required=False)
     parser.add_argument("--analyze", action="store_true", dest="analyze_results", help="Run the results analyzer after each spray interval. False positives are more likely", required=False)
+    parser.add_argument("--jitter", type=int, dest="jitter", help="Jitter time between requests in seconds.", required=False)
+    parser.add_argument("--jitter_min", type=int, dest="jitter_min", help="Minimum time between requests in seconds.", required=False)
```
Comment (Tw1sm) [inline spraycharles.py:38]: Change flag from `--jitter_min` to `--jitter-min` and update a few references

File: README.md [hunk 2]
```diff
@@ -47,6 +64,9 @@ optional arguments:
   -d DOMAIN, --domain DOMAIN
                         HTTP: Prepend DOMAIN\ to usernames. SMB: Specify
                         domain for smb connection
+  --jitter JITTER         Jitter time between requests in seconds.
+  --jitter_min JITTER_MIN Minimum time between requests in seconds.
```
Comment (Tw1sm) [inline README.md:68]: Change flag from `--jitter_min` to `--jitter-min`

File: spraycharles.py (hunk 1 above)
Comment (Tw1sm) [reply spraycharles.py:38]: Around line `190` could you also add an info line that shows the jitter settings? Such as `[*] Jitter: 5-10 seconds between every attempt`

**Review by Tw1sm — CHANGES_REQUESTED** [review]
Hey man, finally got around to reviewing your PR, sorry for the delay. Gave each of the changes a quick test and didn't encounter any issues. Overall seems like really good, functional improvements. I just left a couple comments hoping you could change up a flag name and print the jitter settings to the screen, if provided, with the other settings. If you can make those changes when you have time, I'm good with merging these into master.

Going along with the new Docker option, what's the best way to specify user/pass files when using it through a docker container? I'm a noob Docker user and I got the container working but obviously couldn't reference files outside of the container. I know you can copy the files in, but was wondering if there's a better way.

**Comment by puzzlepeaches** [comment]
Hey, I have been super busy as well, so no worries at all! Used both the ADFS and EWS modules on an engagement recently with great success! 

I made the requested changes and provided some additional direction for using the Dockerfile. On a completely unrelated note, I also changed the Ews.py module to better mirror EWS auth requests made by the Mac OS Mail application. 

I think you should be able to see the changes and merge them to master, but let me know if I need to do something else. I am a bit of a git noob sometimes.  Have a great weekend!

**Review by Tw1sm — APPROVED** [review]

**Comment by Tw1sm** [comment]
Merged - thanks a lot for the contribution!

**Pull request merged**